File: examples/performance_examples.js
/**
 * Performance Optimization Examples - JavaScript
 * 
 * This file demonstrates common performance issues and their optimized solutions.
 */

// ============================================================================
// Example 1: Array Operations - Nested Loops vs Set Lookup
// ============================================================================

console.log('=== Example 1: Array Operations ===\n');

// Inefficient: O(n²) complexity
function findCommonElementsSlow(arr1, arr2) {
  const common = [];
  for (let i = 0; i < arr1.length; i++) {
    for (let j = 0; j < arr2.length; j++) {
      if (arr1[i] === arr2[j]) {
        common.push(arr1[i]);
        break;
      }
    }
  }
  return common;
}

// Optimized: O(n) complexity
function findCommonElementsFast(arr1, arr2) {
  const set2 = new Set(arr2);
  return arr1.filter(item => set2.has(item));
}

// Benchmark
const testArr1 = Array.from({ length: 1000 }, (_, i) => i);
const testArr2 = Array.from({ length: 1000 }, (_, i) => i * 2);

console.time('Slow (nested loops)');
findCommonElementsSlow(testArr1, testArr2);
console.timeEnd('Slow (nested loops)');

console.time('Fast (Set lookup)');
findCommonElementsFast(testArr1, testArr2);
console.timeEnd('Fast (Set lookup)');

console.log('\n');

// ============================================================================
// Example 2: String Concatenation
// ============================================================================

console.log('=== Example 2: String Concatenation ===\n');

// Inefficient: Creates new string each iteration
function buildStringSlow(parts) {
  let result = '';
  for (let part of parts) {
    result += part + ' ';
  }
  return result;
}

// Optimized: Uses array join
function buildStringFast(parts) {
  return parts.join(' ');
}

const testParts = Array.from({ length: 1000 }, (_, i) => `word${i}`);

console.time('Slow (concatenation)');
buildStringSlow(testParts);
console.timeEnd('Slow (concatenation)');

console.time('Fast (array join)');
buildStringFast(testParts);
console.timeEnd('Fast (array join)');

console.log('\n');

// ============================================================================
// Example 3: Object Property Access
// ============================================================================

console.log('=== Example 3: Object Property Access ===\n');

// Inefficient: Repeated property access
function processDataSlow(data) {
  let sum = 0;
  for (let i = 0; i < data.items.length; i++) {
    sum += data.items[i].value * data.multiplier;
  }
  return sum;
}

// Optimized: Cache frequently accessed properties
function processDataFast(data) {
  const items = data.items;
  const multiplier = data.multiplier;
  const length = items.length;
  let sum = 0;
  
  for (let i = 0; i < length; i++) {
    sum += items[i].value * multiplier;
  }
  return sum;
}

const testData = {
  items: Array.from({ length: 10000 }, (_, i) => ({ value: i })),
  multiplier: 2
};

console.time('Slow (repeated access)');
processDataSlow(testData);
console.timeEnd('Slow (repeated access)');

console.time('Fast (cached properties)');
processDataFast(testData);
console.timeEnd('Fast (cached properties)');

console.log('\n');

// ============================================================================
// Example 4: Memoization
// ============================================================================

console.log('=== Example 4: Memoization ===\n');

// Inefficient: Recalculates values
function fibonacciSlow(n) {
  if (n <= 1) return n;
  return fibonacciSlow(n - 1) + fibonacciSlow(n - 2);
}

// Optimized: Memoized version
const fibonacciFast = (() => {
  const cache = new Map();
  return function fib(n) {
    if (n <= 1) return n;
    if (cache.has(n)) return cache.get(n);
    const result = fib(n - 1) + fib(n - 2);
    cache.set(n, result);
    return result;
  };
})();

console.time('Slow (no memoization) - fib(35)');
fibonacciSlow(35);
console.timeEnd('Slow (no memoization) - fib(35)');

console.time('Fast (memoized) - fib(35)');
fibonacciFast(35);
console.timeEnd('Fast (memoized) - fib(35)');

console.log('\n');

// ============================================================================
// Example 5: Array Methods
// ============================================================================

console.log('=== Example 5: Array Methods ===\n');

// Inefficient: Multiple array iterations
function processArraySlow(arr) {
  const filtered = arr.filter(x => x > 10);
  const doubled = filtered.map(x => x * 2);
  const sum = doubled.reduce((acc, x) => acc + x, 0);
  return sum;
}

// Optimized: Single iteration
function processArrayFast(arr) {
  return arr.reduce((acc, x) => {
    if (x > 10) {
      return acc + (x * 2);
    }
    return acc;
  }, 0);
}

const testArray = Array.from({ length: 100000 }, (_, i) => i);

console.time('Slow (multiple iterations)');
processArraySlow(testArray);
console.timeEnd('Slow (multiple iterations)');

console.time('Fast (single iteration)');
processArrayFast(testArray);
console.timeEnd('Fast (single iteration)');

console.log('\n');

// ============================================================================
// Example 6: Debouncing for Event Handlers
// ============================================================================

console.log('=== Example 6: Debouncing ===\n');

// Inefficient: Executes on every event
function onInputSlow(event) {
  console.log('Processing:', event.target.value);
  // Expensive operation here
}

// Optimized: Debounced version
function debounce(func, delay) {
  let timeoutId;
  return function (...args) {
    clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func.apply(this, args), delay);
  };
}

const onInputFast = debounce((event) => {
  console.log('Processing:', event.target.value);
  // Expensive operation here
}, 300);

console.log('Debouncing prevents excessive function calls during rapid events');
console.log('Example: Input events, scroll events, resize events\n');

// ============================================================================
// Example 7: Early Return / Short Circuit
// ============================================================================

console.log('=== Example 7: Early Return ===\n');

// Inefficient: Checks all conditions
function validateUserSlow(user) {
  let valid = true;
  
  if (!user) valid = false;
  if (user && !user.name) valid = false;
  if (user && !user.email) valid = false;
  if (user && user.email && !user.email.includes('@')) valid = false;
  
  return valid;
}

// Optimized: Early return
function validateUserFast(user) {
  if (!user) return false;
  if (!user.name) return false;
  if (!user.email) return false;
  if (!user.email.includes('@')) return false;
  
  return true;
}

console.log('Early return pattern avoids unnecessary checks');
console.log('Especially important when checks are expensive\n');

// ============================================================================
// Performance Tips Summary
// ============================================================================

console.log('=== Performance Tips Summary ===\n');
console.log('1. Use appropriate data structures (Set, Map) for lookups');
console.log('2. Avoid string concatenation in loops - use array join');
console.log('3. Cache frequently accessed properties');
console.log('4. Implement memoization for expensive recursive functions');
console.log('5. Minimize array iterations - combine operations when possible');
console.log('6. Debounce expensive event handlers');
console.log('7. Use early returns to avoid unnecessary checks');
console.log('8. Always benchmark to validate optimizations');

File: examples/performance_examples.py
"""
Performance Optimization Examples - Python

This file demonstrates common performance issues and their optimized solutions.
"""

import time
from functools import lru_cache
from typing import List, Set

# ============================================================================
# Example 1: List Operations - Nested Loops vs Set Lookup
# ============================================================================

print('=== Example 1: List Operations ===\n')

# Inefficient: O(n²) complexity
def find_common_elements_slow(list1: List[int], list2: List[int]) -> List[int]:
    common = []
    for item1 in list1:
        for item2 in list2:
            if item1 == item2:
                common.append(item1)
                break
    return common

# Optimized: O(n) complexity using set
def find_common_elements_fast(list1: List[int], list2: List[int]) -> List[int]:
    set2 = set(list2)
    return [item for item in list1 if item in set2]

# Benchmark
test_list1 = list(range(1000))
test_list2 = list(range(0, 2000, 2))

start = time.perf_counter()
result1 = find_common_elements_slow(test_list1, test_list2)
print(f'Slow (nested loops): {(time.perf_counter() - start) * 1000:.2f}ms')

start = time.perf_counter()
result2 = find_common_elements_fast(test_list1, test_list2)
print(f'Fast (set lookup): {(time.perf_counter() - start) * 1000:.2f}ms\n')

# ============================================================================
# Example 2: String Concatenation
# ============================================================================

print('=== Example 2: String Concatenation ===\n')

# Inefficient: Creates new string each iteration
def build_string_slow(parts: List[str]) -> str:
    result = ''
    for part in parts:
        result += part + ' '
    return result

# Optimized: Uses join
def build_string_fast(parts: List[str]) -> str:
    return ' '.join(parts)

test_parts = [f'word{i}' for i in range(1000)]

start = time.perf_counter()
result1 = build_string_slow(test_parts)
print(f'Slow (concatenation): {(time.perf_counter() - start) * 1000:.2f}ms')

start = time.perf_counter()
result2 = build_string_fast(test_parts)
print(f'Fast (join): {(time.perf_counter() - start) * 1000:.2f}ms\n')

# ============================================================================
# Example 3: List Comprehension vs Loop
# ============================================================================

print('=== Example 3: List Comprehension ===\n')

# Inefficient: Using append in loop
def square_numbers_slow(numbers: List[int]) -> List[int]:
    result = []
    for num in numbers:
        result.append(num ** 2)
    return result

# Optimized: List comprehension
def square_numbers_fast(numbers: List[int]) -> List[int]:
    return [num ** 2 for num in numbers]

test_numbers = list(range(10000))

start = time.perf_counter()
result1 = square_numbers_slow(test_numbers)
print(f'Slow (loop with append): {(time.perf_counter() - start) * 1000:.2f}ms')

start = time.perf_counter()
result2 = square_numbers_fast(test_numbers)
print(f'Fast (list comprehension): {(time.perf_counter() - start) * 1000:.2f}ms\n')

# ============================================================================
# Example 4: Memoization with lru_cache
# ============================================================================

print('=== Example 4: Memoization ===\n')

# Inefficient: Recalculates values
def fibonacci_slow(n: int) -> int:
    if n <= 1:
        return n
    return fibonacci_slow(n - 1) + fibonacci_slow(n - 2)

# Optimized: Using lru_cache decorator
@lru_cache(maxsize=None)
def fibonacci_fast(n: int) -> int:
    if n <= 1:
        return n
    return fibonacci_fast(n - 1) + fibonacci_fast(n - 2)

start = time.perf_counter()
result1 = fibonacci_slow(30)
print(f'Slow (no memoization) - fib(30): {(time.perf_counter() - start) * 1000:.2f}ms')

start = time.perf_counter()
result2 = fibonacci_fast(30)
print(f'Fast (memoized) - fib(30): {(time.perf_counter() - start) * 1000:.2f}ms\n')

# ============================================================================
# Example 5: Generator vs List
# ============================================================================

print('=== Example 5: Generator vs List ===\n')

# Inefficient: Loads all data into memory
def process_large_dataset_slow(n: int) -> int:
    data = [i ** 2 for i in range(n)]
    return sum(data)

# Optimized: Uses generator
def process_large_dataset_fast(n: int) -> int:
    return sum(i ** 2 for i in range(n))

n = 1000000

start = time.perf_counter()
result1 = process_large_dataset_slow(n)
print(f'Slow (list): {(time.perf_counter() - start) * 1000:.2f}ms')

start = time.perf_counter()
result2 = process_large_dataset_fast(n)
print(f'Fast (generator): {(time.perf_counter() - start) * 1000:.2f}ms\n')

# ============================================================================
# Example 6: Dictionary Lookup vs List Search
# ============================================================================

print('=== Example 6: Dictionary Lookup ===\n')

# Inefficient: Linear search in list
def has_permission_slow(user_id: int, allowed_users: List[int]) -> bool:
    return user_id in allowed_users

# Optimized: Hash lookup in set
def has_permission_fast(user_id: int, allowed_users: Set[int]) -> bool:
    return user_id in allowed_users

test_list = list(range(10000))
test_set = set(range(10000))
search_id = 9999

start = time.perf_counter()
for _ in range(1000):
    has_permission_slow(search_id, test_list)
print(f'Slow (list search): {(time.perf_counter() - start) * 1000:.2f}ms')

start = time.perf_counter()
for _ in range(1000):
    has_permission_fast(search_id, test_set)
print(f'Fast (set lookup): {(time.perf_counter() - start) * 1000:.2f}ms\n')

# ============================================================================
# Example 7: Using any() and all() for Short-Circuit Evaluation
# ============================================================================

print('=== Example 7: Short-Circuit Evaluation ===\n')

# Inefficient: Checks all elements
def has_negative_slow(numbers: List[int]) -> bool:
    result = False
    for num in numbers:
        if num < 0:
            result = True
    return result

# Optimized: Uses any() with generator
def has_negative_fast(numbers: List[int]) -> bool:
    return any(num < 0 for num in numbers)

test_numbers = list(range(10000)) + [-1]

start = time.perf_counter()
result1 = has_negative_slow(test_numbers)
print(f'Slow (no short-circuit): {(time.perf_counter() - start) * 1000:.2f}ms')

start = time.perf_counter()
result2 = has_negative_fast(test_numbers)
print(f'Fast (any with short-circuit): {(time.perf_counter() - start) * 1000:.2f}ms\n')

# ============================================================================
# Example 8: Avoid Global Variable Lookups
# ============================================================================

print('=== Example 8: Local vs Global Variables ===\n')

MULTIPLIER = 2

# Inefficient: Global variable lookup in tight loop
def compute_slow(numbers: List[int]) -> List[int]:
    return [num * MULTIPLIER for num in numbers]

# Optimized: Cache global as local
def compute_fast(numbers: List[int]) -> List[int]:
    multiplier = MULTIPLIER  # Local variable
    return [num * multiplier for num in numbers]

test_numbers = list(range(100000))

start = time.perf_counter()
result1 = compute_slow(test_numbers)
print(f'Slow (global lookup): {(time.perf_counter() - start) * 1000:.2f}ms')

start = time.perf_counter()
result2 = compute_fast(test_numbers)
print(f'Fast (local variable): {(time.perf_counter() - start) * 1000:.2f}ms\n')

# ============================================================================
# Performance Tips Summary
# ============================================================================

print('=== Performance Tips Summary ===\n')
print('1. Use set/dict for membership testing instead of list')
print('2. Use join() for string concatenation, not +=')
print('3. Prefer list comprehensions over loops with append')
print('4. Use @lru_cache for expensive recursive functions')
print('5. Use generators for large datasets to save memory')
print('6. Use any()/all() for short-circuit evaluation')
print('7. Cache global variables as local in tight loops')
print('8. Always profile with timeit or cProfile before optimizing')
